File: spring-boot-starter-breuninger-validation/src/main/java/com/breuninger/boot/validation/validators/EnumValidator.java
package com.breuninger.boot.validation.validators;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class EnumValidator implements ConstraintValidator<IsEnum, String> {

  private Set<String> availableEnumNames;
  private boolean ignoreCase;
  private boolean allowNull;

  @Override
  public void initialize(final IsEnum annotation) {
    final Class<? extends Enum<?>> enumClass = annotation.enumClass();
    availableEnumNames = Stream.of(enumClass.getEnumConstants()).map(Enum::name).collect(Collectors.toSet());
    ignoreCase = annotation.ignoreCase();
    allowNull = annotation.allowNull();
  }

  @Override
  public boolean isValid(final String value, final ConstraintValidatorContext context) {
    if (value == null) {
      return allowNull;
    } else {
      return availableEnumNames.stream().anyMatch(o -> {
        if (ignoreCase) {
          return o.equalsIgnoreCase(value);
        } else {
          return o.equals(value);
        }
      });
    }
  }
}

File: validation/src/main/kotlin/com/breuninger/boot/validation/constraintvalidators/EnumValidator.kt
package com.breuninger.boot.validation.constraintvalidators

import com.breuninger.boot.validation.constraints.IsEnum
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class EnumValidator : ConstraintValidator<IsEnum, String> {

  private var availableEnumNames: Set<String> = HashSet()
  private var ignoreCase = false
  private var allowNull = false

  override fun initialize(annotation: IsEnum) {
    availableEnumNames = annotation.enumClass.java.enumConstants.map { it.name }.toSet()
    ignoreCase = annotation.ignoreCase
    allowNull = annotation.allowNull
  }

  override fun isValid(value: String?, context: ConstraintValidatorContext) = value?.let {
    availableEnumNames.any {
      return if (ignoreCase) {
        it.equals(value, ignoreCase = true)
      } else {
        it == value
      }
    }
  } ?: allowNull
}
